docs/architecture-decision-records/20220121-conduit-plugin-architecture.md

```diff
@@ -0,0 +1,322 @@
+# Conduit Plugin Architecture
+
+## Summary
+
+This doc describes the design for the conduit plugin architecture. It spans from how Conduit communicates with built-in
+and standalone plugins, where and how the interface is defined, how the plugin SDK is defined.
+
+## Context
+
+We want to provide a good development experience to plugin developers, we also want to be able to ship Conduit with real
+built-in plugins (compiled into the Conduit binary), we want to make it as easy as possible to write plugins in any
+programming language and we want the SDK to be decoupled from Conduit and be able to change without changing Conduit
+itself.
+
+## Decision
+
+The decision can be broken up into 4 parts:
+
+- We defined the plugin interface in gRPC
+- We defined the interface used internally in Conduit to communicate with plugins
+- We introduced plugin registries to support built-in and standalone plugins
+- We defined the interface for the plugin SDK
+
+### Conduit plugin interface (gRPC)
+
+The Conduit plugin interface is defined in gRPC, is standalone (does not depend on Conduit or the SDK) and lives
+in https://github.com/conduitio/conduit-plugin. This repository is the only common thing between Conduit and a plugin,
+meaning that Conduit uses the interface to interact with plugins and plugins implement the interface.
+
+The proto files define the messages and gRPC interface that needs to be implemented by the connector plugin. The
+interface is uploaded to the [buf schema registry](https://buf.build/conduitio/conduit-plugin), which allows us to
+easily generate the code in any programming language. For Go we
+use [remote generation](https://docs.buf.build/bsr/remote-generation/overview) so we don’t have to manually generate any
+code.
+
+Alongside the proto files, the repository contains helper structs and methods that hide the gRPC implementation from the
+user without changing any semantics. No other code needs to know that the communication happens through gRPC (i.e.
+neither Conduit nor the plugin SDK).
```

Comment:
My bad! 🤦  You're right, Conduit defines its own gRPC client, in the text I'm talking about the thin layer in [conduit-plugin](https://github.com/conduitio/conduit-plugin) which only hides the gRPC implementation from the SDK. I decided to put the gRPC client code into Conduit since it is the only piece of code that needs the client, we can keep it close to where it's used and modify it without worrying about breaking the interface for anyone else. Also, it's one less layer to worry about.

Reply:
Sounds very reasonable to me. 👍